### DISCUSSION TITLE
Suggestion: update the "-s" flag default to 1

### DISCUSSION THREAD

**Discussion by gregbenz** [google/libultrahdr#246] [Ideas]
The user is in control of gain map down sampling via the "**-s**" flag in the demo app. However, I recommend updating the default to 1 (vs the current 4) to ensure the default behavior is correct most of the time.

For images where quality is important or the resolution is low, a 1:1 sampling is required. A lower 1:2 sampling rate can be a reasonable choice to reduce file size in many cases, but quality will be degraded. The default 1:4 is almost never ideal as it will frequently cause quality problems in high frequency detail such as edges.

**Comment by cmahnke** [comment]
Well, I guess the current setting is due to the increased file size if there is no subsampling of the gainmap. Keep in mind that this code had more quite arbitrary looking limitations (like the max size) that were introduced by the primary use case of mobile phones.

Maybe it's an workable approach to consider profiles? Like `photography` which would set 1:1 sampling, (this one is made up) `web` with 1:2 and `mobile` with 1:4 (current default).
Additionally those profiles could also imply more settings:
| profile     | encoding preset (`-D`) | quality factor (-Q) |
|----------|------------------------|-------------------|
| `photo`   | 1                                      | 100                         |
| `mobile` | 0                                      | 85                           |

**Reply by ram-mohan** [reply]
@cmahnke @gregbenz The max size (UHDR_MAX_DIMENSION) by default is configured to 8192, making the max supported resolution to 8192x8192 capable of covering most mobile camera capture resolutions. This can be modified to the choice desired at config time using `cmake -DUHDR_MAX_DIMENSION=value `.  

The initial revisions of the library did not support features multi-channel gainmap encoding, encoding speed preset. These were added later. The default value of these features are chosen such that they are at parity with base version. The gainmap scale factor defaulting to 4 can be an issue as this might result in missing high frequency details like small bright points of light. This needs revision. Choice of quality factor is a subjective preference and best left to user.

In the newer version of the library, there are two settings we would like to modify. The gainmap scale factor defaulting to '1' and enabling multi channel gainmap by default, both orienting towards better encoding quality.

**Reply by cmahnke** [reply]
@ram-mohan: Would a gainmap scale factor of 1 allow for uneven dimensions? Or why has it to be even currently?

**Reply by gregbenz** [reply]
My recommendation would be that the default encoding settings should be based on recommendations for reasonably high quality images in order to ensure good results if the end user does not understand the tool.

From there, app flags may be used to select lower quality, and recommended settings may be shared in this forum or others - such as suggestions for various quality vs size objectives.

Optional profiles may be useful, but I recommend they are based on quality objectives rather than assuming anything about the device.

For example, I do not believe there is a single optimal group of settings for mobile. A 1:2 gain map may be ideal for native capture where the map is low quality (auto generated) and size is a concern when capturing many images. However, it may be ideal for the same mobile phone to produce a higher quality map if an image is edited by a user on the same device using a RAW capture instead of a native gain map as input. 

Naming the profile for the quality / encoding objectives rather than the host device is probably more clear to ensure proper use.

**Reply by ram-mohan** [reply]
@cmahnke This comes down to how the library is handling the 420 subsampling format. The library supports input color formats UHDR_IMG_FMT_24bppYCbCrP010, UHDR_IMG_FMT_12bppYCbCr420. Both these formats have chroma planes subsampled by 2, horizontally and vertically. Each chroma pixel corresponds to 4 luma pixels. Library holds this attribute too tightly and evaluates chroma plane dimensions as half of luma plane dimensions (round_down(w / 2) and round_down(h/2)). It is possible to have 420 subsampled images with odd dimensions where the edge chroma pixels correspond to 1 or 2 luma pixels but this was not handled. Once the evaluations in the library are modified to (round_up(w / 2) and round_up(h/2)) in subsampling formats, this restriction can be lifted. For rgb inputs, this restriction should not be enforced. But we are in the process of adding support for odd dimensions for all formats, so no changes were made specific to rgb.

**Reply by cmahnke** [reply]
@gregbenz: You're right about naming, I totally agree. 
Just for clarification, we are talking about the app here, aren't we? Since it certainly won't be used itself in the context of a mobile phone (instead a implementing library would use the library) primary target audiences are certainly professionals like you.

**Reply by cmahnke** [reply]
@ram-mohan: Thanks for pointing this out! Should we add the support of uneven dimensions to the issue tracker?

**Reply by gregbenz** [reply]
I'm thinking mostly of the app. But for either the app or library: I would suggest any default/recommended settings optimized for reasonably high quality, and any alternatives should be named for the encoding objective to ensure clarity as to which option set is best suited for a given use.

**Reply by ram-mohan** [reply]
@cmahnke yes please, thank you.

**Comment by ram-mohan** [comment]
The desired changes are merged to main. This can be marked as resolved.
